Hw spring boot practice3

Petro-Smoliar opened the pull request.

Petro-Smoliar committed:
Develop shopping cart functionality for users in a Spring Boot project.
Introduce entities like ShoppingCart and CartItem, integrating Liquibase for database management.
Implement a soft delete approach and ensure pagination, sorting, and Swagger documentation for all controllers.
Create endpoints for retrieving, adding, updating quantities, and removing books from the shopping cart.
Enforce security measures, distinguishing access based on user roles (USER, ADMIN).
Fulfill specific endpoint requirements for both authenticated and non-authenticated users.

Petro-Smoliar committed:
Add Order and OrderItem entities to represent user orders and their items.
Implement Liquibase for database schema management.
Implement soft delete approach for data integrity.
Add endpoints for users to place orders, view order history, and retrieve order items.
Add admin endpoints for updating order status.
Implement Pagination, Sorting, and Swagger for all relevant controllers.
Ensure consistency with existing User, Role, Book, Category, and ShoppingCart endpoints.

okuzan commented on src/main/java/com/example/book/store/controller/OrderController.java:
```diff
@@ -0,0 +1,72 @@
+            description = "Allows a user to place an order by sending a "
+                              + "POST request to /api/orders."
+    )
+    @PreAuthorize("hasRole('ROLE_USER')")
```
this setup won't allow Admins to access endpoints annotated with ` @PreAuthorize("hasRole('ROLE_USER')")`. let's remove them. They are configured accessible only for authenticated users in Security Config anyway

okuzan commented on src/main/java/com/example/book/store/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,73 @@
+@Tag(name = "Shopping cart management", description = "Endpoints for managing shopping cart")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/api")
```
make api project wide path prefix

okuzan commented on src/main/java/com/example/book/store/dto/order/OrderRequestStatusDto.java:
```diff
@@ -0,0 +1,10 @@
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+
+public record OrderRequestStatusDto(
```
where is it used?

okuzan commented on src/main/java/com/example/book/store/model/Order.java:
```diff
@@ -0,0 +1,61 @@
+
+@Entity
+@Data
+@Where(clause = "is_deleted = false")
```
```suggestion
@SQLDelete(sql = "UPDATE orders SET is_deleted = TRUE WHERE id = ?")
@Where(clause = "is_deleted = FALSE")
```

okuzan commented on src/main/java/com/example/book/store/model/Order.java:
```diff
@@ -0,0 +1,61 @@
+        COMPLETED
+    }
+}
+
```
```suggestion
```

okuzan commented on src/main/resources/db/changelog/change/12-create-order_items-table.yaml:
```diff
@@ -0,0 +1,35 @@
+                  name: price
+                  type: decimal
+                  constraints:
+                    nullable: false
```
make sure each file ends with  a new line

okuzan commented on src/main/resources/db/changelog/change/11-create-orders-table.yaml:
```diff
@@ -0,0 +1,45 @@
+                  name: is_deleted
+                  type: boolean
+                  constraints:
+                    nullable: false
```
same

okuzan commented on src/main/resources/db/changelog/change/10-create-cart_items-table.yaml:
```diff
@@ -0,0 +1,35 @@
+                  name: is_deleted
+                  type: boolean
+                  constraints:
+                    nullable: false
```
same

Petro-Smoliar committed:
Add Order and OrderItem entities to represent user orders and their items.
Implement Liquibase for database schema management.
Implement soft delete approach for data integrity.
Add endpoints for users to place orders, view order history, and retrieve order items.
Add admin endpoints for updating order status.
Implement Pagination, Sorting, and Swagger for all relevant controllers.
Ensure consistency with existing User, Role, Book, Category, and ShoppingCart endpoints.
